### ISSUE TITLE
Add support for asynchronous invallidation of queries

### ISSUE THREAD

**Issue by zserik** [salsa-rs/salsa#215]
[some context](https://www.reddit.com/r/rust/comments/fyq73s/ondemand_salsa_filesystem_inputs_with/), [fileqry repo](https://github.com/zserik/fileqry)
> I stumbled across the Salsa book while investigating potential usages of salsa in my project and found the on-demand pattern quite interesting.

The on-demand pattern example uses `query_mut`, which requires an `&mut impl salsa::Database`. Obviously, queries can't be invalidated while other queries are active. Then I thought about wrapping the database object in an `RwLock`, but that requires that the database and `Runtime` is `Sync`, which isn't the case.

I currently solved this using a separate `process_events` function, which utilizes `crossbeam-channel` to receive information about invalidated queries, but this approach doesn't really scale. In larger interfaces, the `process_events` function should be called at most entry points into the interface.

A possible solution would be: adding a way to invalidate queries asynchronously (e.g. invalidation asap, but not now); Currently running queries which (directly or indirectly) depend on the invalides queries should be canceled, then the invalidation itself should be applied (after the affected queries returned/unwinded).

**Comment by cormacrelf** [comment]
I’m pretty sure what you’re describing already exists, documented in the query_mut method.

https://docs.rs/salsa/0.14.1/salsa/trait.Database.html#method.query_mut

If all the queries check for the cancellation flag frequently enough, then you can ensure the mutating thread never blocks for a long time. You choose if you want to use the panic/unwind/catch method or use Result everywhere. You do not need an RwLock over the whole database because this is already built-in, as the snapshot function. You send a snapshot to another thread, you run queries on it, but mutate the original. Because “mutate the original” is blocking, you can instead send all mutations over a channel to a thread that simply applies them in a loop. The more the queries check the flag, the less latency you get from mutations. Salsa generates the required mutation objects for you, I think you can probably just send them and apply them generically, although closures would be fine. And if you want it to appear async, send an async oneshot channel Tx as well, fire it when done.

**Comment by zserik** [comment]
ok, so the intended method is: having one owning thread, which accumulates all mutations and invalidations; the other threads send the update information to that one.

**Comment by cormacrelf** [comment]
I should mention that you can't hold on to snapshots in other threads without blocking mutations, so they need to be created and destroyed often. Creation happens wherever the database is owned, so if you have an owning thread accumulating mutations, it would also have to be the starting point for a query. With cancellations & retries this is tricky, because you need to:

```rust
use salsa::{Snapshot, Database};

// these types are extremely kludgey but you get the gist
struct ClosureQuery<T: 'static> {
    closure: Box<dyn Clone + Fn(Snapshot<MyDatabase>) -> Result<T, QueryCancelled>>,
    done: Sender<T>,
    reschedule: Sender<Msg>,
}

trait Query { fn execute(&self, snap: Snapshot<MyDatabase>); }

impl<T> Query for ClosureQuery<T> {
    fn execute(&self, snap: Snapshot<MyDatabase>) {
        let ClosureQuery { closure, done, reschedule } = *self;
        // you probably want {catch_unwind and turn the panic value check into Err/re-panicking}
        // in this function instead of every single closure
        match closure(snap) {
            Ok(x) => done.send(x).unwrap(),
            // the query finished early to make way for a mutation, reschedule it
            Err(QueryCancelled) => {
                 let retry = reschedule.clone();
                 // very kludgey reallocation but you get it
                 retry.send(Msg::Query(Box::new(ClosureQuery { closure, done, reschedule }))).unwrap();
            }
        };
    }
}

enum Msg<T> {
    Query(Box<dyn Query>),
    Mutation(Box<dyn FnOnce(&mut MyDatabase) -> ()>),
}

let mut database; // owned or at least &mut borrowed here
loop {
    match channel.recv()? {
        q @ Msg::Query(query) => {
            // create snap outside the threadpool so we can move it in without moving
            // or borrowing the database itself
            let snap = database.snap();
            some_threadpool.spawn(move || {
                query.execute(snap);
                // snapshot dropped, i.e. rwlock read-lock released, here
            });
        }
        Msg::Mutation(mutation_closure) => mutation_closure(&mut database),
    }
}
```

In summary:

1. accept queries and mutations over a channel
2. dispatch a thread for each query (or using a threadpool etc). This thread should have some way of replying to the request (e.g. oneshot channel tx) and re-enqueueing cancelled ones (e.g. a tx handle to the mpsc used in (1))
3. run mutations in place

At the end of the day, your invalidations of the cached file reads are just mutations that cause other queries to be recomputed when they are next run. Salsa lets you additionally cause other queries to stop executing if, due to fs changes, you know they're already out of date. For this to be useful you need to be running queries in parallel; my pseudocode merely schedules queries in parallel with snapshots while accounting for the retry logic. Without parallel queries operating on snapshots, you may have a whole thread for fs events, but you won't be able to immediately execute the invalidations (ie mutations) and cause it to block, because no parallel queries implies you're taking &MyDatabase borrows for the queries, and must interleave the mutations. (This is your process_events in a nutshell.)
